README.md
# StrictOpenapi3

Stricted OpenApi3 implementation for gourmets.

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'strict_openapi3'
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install strict_openapi3

## Usage

``` ruby
validator = StrictOpenapi3::Validator.new("http://some-site.com/api.yml")
begin
  validator.validate_request(method, path, params, content_type, body)
  validator.validate_response(method, path, status, content_type, body)
rescue StrictOpenapi3::Error => e
end
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/uchiru/strict_openapi3. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.

## License

The gem is available as open source under the terms of the [MIT License](http://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the StrictOpenapi3 project’s codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/[USERNAME]/strict_openapi3/blob/master/CODE_OF_CONDUCT.md).


CHANGELOG.md
* Initial release
* Remove `format` property for string item
* Add `pattern` property for string item
* Fix nullable for several refs


spec/spec_helper.rb
require "strict_openapi3"

module Helpers
  # File activesupport/lib/active_support/core_ext/kernel/reporting.rb, line 51
  def silence_stream(stream)
    old_stream = stream.dup
    stream.reopen(RbConfig::CONFIG['host_os'] =~ /mswin|mingw/ ? 'NUL:' : '/dev/null')
    stream.sync = true
    yield
  ensure
    stream.reopen(old_stream)
    old_stream.close
  end
end

RSpec.configure do |config|
  config.expect_with :rspec do |expectations|
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  config.mock_with :rspec do |mocks|
    mocks.verify_partial_doubles = true
  end

  config.shared_context_metadata_behavior = :apply_to_host_groups

  config.include Helpers
end


spec/schema/parser_spec.rb
describe StrictOpenapi3::Parser do
  it "failed on missed required key" do
    expect do
      StrictOpenapi3::Parser.new({
        "openapi" => "3.0.0",
        "paths" => {},
      }).to_schema!
    end.to raise_error('Broken schema: "/" missed required keys ["info"]')
  end

  it "failed on unknown key" do
    expect do
      StrictOpenapi3::Parser.new({
        "openapi" => "3.0.0",
        "info" => {},
        "paths" => {},
        "unknown_key" => "yo"
      }).to_schema!
    end.to raise_error('Broken schema: "/" unknown keys ["unknown_key"]')
  end

  it "failed on wrong openapi" do
    expect do
      StrictOpenapi3::Parser.new({
        "openapi" => "2.0.0",
        "info" => {},
        "paths" => {},
      }).to_schema!
    end.to raise_error('Broken schema: "/openapi" expect "3.0.0" but get "2.0.0"')
  end

  describe "wrong component" do
    it "missed properties for object" do
      expect do
        StrictOpenapi3::Parser.new({
          "openapi" => "3.0.0",
          "info" => {"version" => "0.0.0", "title" => "test schema"},
          "components" => {"schemas" => {"Some" => {"type" => "object"}}},
          "paths" => {},
        }).to_schema!
      end.to raise_error('Broken schema: "/components:Some[object]" missed required keys ["properties"]')
    end

    it "unknown attribute for object" do
      expect do
        StrictOpenapi3::Parser.new({
          "openapi" => "3.0.0",
          "info" => {"version" => "0.0.0", "title" => "test schema"},
          "components" => {"schemas" => {"Some" => {"type" => "object", "properties" => {}, "unknown_attribute" => {}}}},
          "paths" => {},
        }).to_schema!
      end.to raise_error('Broken schema: "/components:Some" unknown keys ["unknown_attribute"]')
    end

    it "broken required for object" do
      expect do
        StrictOpenapi3::Parser.new({
          "openapi" => "3.0.0",
          "info" => {"version" => "0.0.0", "title" => "test schema"},
          "components" => {"schemas" => {"Some" => {"type" => "object", "properties" => {}, "required" => "broken"}}},
          "paths" => {},
        }).to_schema!
      end.to raise_error('Broken schema: "/components:Some:required" should be array')
    end

    it "unknown required for object" do
      expect do
        StrictOpenapi3::Parser.new({
          "openapi" => "3.0.0",
          "info" => {"version" => "0.0.0", "title" => "test schema"},
          "components" => {"schemas" => {"Some" => {"type" => "object", "properties" => {}, "required" => ["some"]}}},
          "paths" => {},
        }).to_schema!
      end.to raise_error('Broken schema: "/components:Some" "["some"]" are required but not descipted in properties section')
    end
  end

  it "failed on wrong path name" do
    [
      "this-is-not-valid-path",
      "/ xxx",
      "/{{abc}}",
      "/{ab}a",
    ].each do |path|
      expect do
        StrictOpenapi3::Parser.new({
          "openapi" => "3.0.0",
          "info" => {"version" => "0.0.0", "title" => "test schema"},
          "paths" => {path => {}},
        }).to_schema!
      end.to raise_error(%(Broken schema: "/paths:#{path}" expect match "#{StrictOpenapi3::Parser::PATH_NAME_RE}"))
    end
  end

  it "failed on wrong verb" do
    expect do
      StrictOpenapi3::Parser.new({
        "openapi" => "3.0.0",
        "info" => {"version" => "0.0.0", "title" => "test schema"},
        "paths" => {"/" => {"doit" => {}}},
      }).to_schema!
    end.to raise_error(%(Broken schema: "/paths:/" unknown "doit", allow to use #{StrictOpenapi3::Parser::VERBS.inspect}))
  end

  it "failed on missed params parameters" do
    expect do
      StrictOpenapi3::Parser.new({
        "openapi" => "3.0.0",
        "info" => {"version" => "0.0.0", "title" => "test schema"},
        "tags" => [{"name" => "tag", "description" => "desc"}],
        "paths" => {"/pets/{id}" => {"get" => {
          "summary" => "get",
          "tags" => ["tag"],
          "responses" => {
            "200" => {}
          }
        }}},
      }).to_schema!
    end.to raise_error(%(Broken schema: "/paths:/pets/{id}:get:parameters" missed path params ["id"]))
  end

  it "failed on unknown path parameters" do
    expect do
      StrictOpenapi3::Parser.new({
        "openapi" => "3.0.0",
        "info" => {"version" => "0.0.0", "title" => "test schema"},
        "tags" => [{"name" => "tag", "description" => "desc"}],
        "paths" => {"/pets" => {"get" => {
          "summary" => "get",
          "tags" => ["tag"],
          "parameters" => [
            {"name" => "unk", "in" => "path", "required" => true, "description" => "unk", "schema" => {"type" => "string"}}
          ],
          "responses" => {
            "200" => {"description" => "suc"}
          }
        }}},
      }).to_schema!
    end.to raise_error(%(Broken schema: "/paths:/pets:get:parameters" unknown path params ["unk"]))
  end

  it "pass valid schema" do
    expected = YAML.load(File.read(File.expand_path("../factories/api_compilated.yaml", __dir__)))
    schema = StrictOpenapi3::Parser.new(YAML.load(File.read(File.expand_path("../factories/api.yaml", __dir__)))).to_schema!
    # File.write(File.expand_path("../factories/api_compilated.yaml", __dir__), YAML.dump(schema))
    expect(schema).to eq(expected)
  end

  it "expects summary for each method" do
    expect do
      StrictOpenapi3::Parser.new({
        "openapi" => "3.0.0",
        "info" => {"version" => "0.0.0", "title" => "test schema"},
        "paths" => {"/pets" => {"get" => {
          "description" => "get",
          "responses" => {
            "200" => {"description" => "succes"}
          }
        }}}
      }).to_schema!
    end.to raise_error('Broken schema: "/paths:/pets:get" missed required keys ["summary", "tags"]')
  end

  it "expects one tag for each method" do
    expect do
      StrictOpenapi3::Parser.new({
        "openapi" => "3.0.0",
        "info" => {"version" => "0.0.0", "title" => "test schema"},
        "paths" => {"/pets" => {"get" => {
          "summary" => "Get pets",
          "tags" => ["tag1", "tag2"],
          "responses" => {
            "200" => {"description" => "succes"}
          }
        }}}
      }).to_schema!
    end.to raise_error('Broken schema: "/paths:/pets:get:tags" should be array with one element')
  end

  it "Each tag should be descripted in head of schema" do
    expect do
      StrictOpenapi3::Parser.new({
        "openapi" => "3.0.0",
        "tags" => [{"name" => "tag1", "description" => "this is tag1"}],
        "info" => {"version" => "0.0.0", "title" => "test schema"},
        "paths" => {"/pets" => {"get" => {
          "summary" => "Get pets",
          "tags" => ["tag2"],
          "responses" => {
            "200" => {"description" => "succes"}
          }
        }}}
      }).to_schema!
    end.to raise_error('Broken schema: "/paths:/pets:get:tags[tag2]" is not descripted in "tags" section of schema')
  end
end


spec/schema/validator_spec.rb
describe StrictOpenapi3::Validator do
  let (:validator) { StrictOpenapi3::Validator.new(File.expand_path("../factories/api.yaml", __dir__)) }

  it "json-schema test" do
    expect do
      JSON::Validator.validate!({
        "type" => "object",
        "required" => ["cmd"],
        "properties" => {
          "cmd" => {"type" => ["string", "null"]}
        }
      }, {"cmd" => nil})
    end.to_not raise_error
  end

  it "json-schema test - 2" do
    expect do
      JSON::Validator.validate!({
        "type" => "object",
        "required" => ["cmd"],
        "properties" => {
          "cmd" => {"type" => ["string", "null"]}
        }
      }, {"cmd" => "cmd"})
    end.to_not raise_error
  end

  describe 'request' do
    it "fails on broken schema" do
      broken = StrictOpenapi3::Validator.new(File.expand_path("../factories/broken.yaml", __dir__))
      expect do
        broken.validate_request("GET", "/stats", {}, "application/json; charset=utf-8", '')
      end.to raise_error(StrictOpenapi3::BrokenSchemaError)
    end

    it 'fails unknown path' do
      expect do
        validator.validate_request("POST", "/unknown", {}, "application/json; charset=utf-8", '')
      end.to raise_error(%(Not found spec for "/unknown"))
    end

    it 'fails unknown verb' do
      expect do
        validator.validate_request("TEAPOT", "/pets", {}, "application/json; charset=utf-8", '')
      end.to raise_error(%(Not found spec for "/pets" "teapot"))
    end

    it 'fails unknown content type' do
      expect do
        expect(validator.validate_request("POST", "/pets", {}, "application/xml", '<pet />'))
      end.to raise_error(%(Not found request body spec for "/pets" "post" "application/xml"))
    end

    it 'fails missed body' do
      expect do
        expect(validator.validate_request("POST", "/pets", {}, "application/json", ''))
      end.to raise_error(%(Request body required but missed for "/pets" "post"))
    end

    it 'fails unrequired body' do
      expect do
        expect(validator.validate_request("GET", "/pets", {"extend" => true}, "application/json", '<body/>'))
      end.to raise_error(%(Request body provided but not defined in schema for "/pets" "get" "application/json"))
    end

    it 'fails broken request schema' do
      expect do
        expect(validator.validate_request("POST", "/pets", {}, "application/json", '{"pet":{}}'))
      end.to raise_error(%(request failed: The property '#/pet' did not contain a required property of 'name'))
    end

    it 'fails broken request schema-2' do
      expect do
        expect(validator.validate_request("POST", "/pets", {}, "application/json", '{"pet":{"name":"bot","age":62}}'))
      end.to raise_error(%(request failed: The property '#/pet' contains additional properties ["age"] outside of the schema when none are allowed))
    end

    it 'fails broken path params' do
      expect do
        expect(validator.validate_request("GET", "/pets/pet", {}, "application/json", ''))
      end.to raise_error(%(Expect integer param "id" but get "pet" for "/pets/pet" "get"))
    end

    it 'passes correct path params' do
      expect do
        expect(validator.validate_request("GET", "/pets/123", {}, "application/json", ''))
      end.to_not raise_error
    end

    it 'fails on missed query param' do
      expect do
        expect(validator.validate_request("GET", "/pets", {}, "application/json", ''))
      end.to raise_error(%(Missed required params ["extend"] for "/pets" "get"))
    end

    it 'fails on broken query param' do
      expect do
        expect(validator.validate_request("GET", "/pets", {"extend" => "abc"}, "application/json", ''))
      end.to raise_error(%(Expect boolean param "extend" but get "abc" for "/pets" "get"))
    end

    it 'fails on unknown query param' do
      expect do
        expect(validator.validate_request("GET", "/pets", {"extend" => true, "some" => 123}, "application/json", ''))
      end.to raise_error(%(Unknown param "some" for "/pets" "get"))
    end

    it 'passes correct query params' do
      expect do
        expect(validator.validate_request("GET", "/pets", {"extend" => false}, "application/json", ''))
      end.to_not raise_error
    end

    it 'passes correct request' do
      expect do
        expect(validator.validate_request("POST", "/pets", {}, "application/json", '{"pet":{"name":"bob"}}'))
      end.to_not raise_error
    end

    it 'passes correct request' do
      expect do
        expect(validator.validate_request("POST", "/pets", {}, "application/json", '{"pet":{"name":"bob"}}'))
      end.to_not raise_error
    end

    it 'failes broken pet name' do
      expect do
        expect(validator.validate_request("POST", "/pets", {}, "application/json", '{"pet":{"name":"0bob"}}'))
      end.to raise_error(%(request failed: The property '#/pet/name' value "0bob" did not match the regex '^[a-z][a-z0-9\\-_]+$'))
    end
  end

  describe 'response' do
    it "fails on broken schema" do
      broken = StrictOpenapi3::Validator.new(File.expand_path("../factories/broken.yaml", __dir__))
      expect do
        broken.validate_response("GET", "/pets", 200, "application/json; charset=utf-8", '{"pets":[]}')
      end.to raise_error(StrictOpenapi3::BrokenSchemaError)
    end

    it 'fails unknown path' do
      expect do
        validator.validate_response("POST", "/unknown", 200, "application/json; charset=utf-8", '')
      end.to raise_error(%(Not found spec for "/unknown"))
    end

    it 'fails unknown verb' do
      expect do
        validator.validate_response("TEAPOT", "/pets", 200, "application/json; charset=utf-8", '')
      end.to raise_error(%(Not found spec for "/pets" "teapot"))
    end

    it 'fails unknown status' do
      expect do
        expect(validator.validate_response("GET", "/pets", 201, "application/json", ''))
      end.to raise_error(%(Not found spec for "/pets" "get" "201"))
    end

    it 'fails unknown content type' do
      expect do
        expect(validator.validate_response("GET", "/pets", 200, "application/xml", '<root/>'))
      end.to raise_error(%(Not found spec for "/pets" "get" "200" "application/xml"))
    end

    it 'fails unknown content type' do
      expect do
        expect(validator.validate_response("GET", "/pets", 200, "application/xml", '<root/>'))
      end.to raise_error(%(Not found spec for "/pets" "get" "200" "application/xml"))
    end

    it 'fails broken body response' do
      expect do
        expect(validator.validate_response("GET", "/pets", 200, "application/json", '{"pets":'))
      end.to raise_error(/schema failed for "\/pets" "get" "200" "application\/json":/)
    end

    it 'fails broken scheman response' do
      expect do
        expect(validator.validate_response("GET", "/pets", 200, "application/json", '{"pets":[{"age":48}]}'))
      end.to raise_error(%(schema failed for "/pets" "get" "200" "application/json": The property '#/pets/0' did not contain a required property of 'id'))
    end

    it 'passes valid response' do
      expect do
        expect(validator.validate_response("GET", "/pets", 200, "application/json", '{"pets":[{"id":123,"name":"bob"}]}'))
      end.to_not raise_error
    end

    it 'passes valid response - 2' do
      expect do
        expect(validator.validate_response("POST", "/pets", 201, "application/json", ''))
      end.to_not raise_error
    end

    it 'passes nullable as null' do
      expect do
        expect(validator.validate_response("GET", "/nullable", 200, "application/json", '{"obj":null}'))
      end.to_not raise_error
    end

    it 'passes nullable as object' do
      expect do
        expect(validator.validate_response("GET", "/nullable", 200, "application/json", '{"obj":{"some":"abc"}}'))
      end.to_not raise_error
    end

    it 'failse nullable as wrong object' do
      expect do
        expect(validator.validate_response("GET", "/nullable", 200, "application/json", '{"obj":{"some":"abc","another":"cde"}}'))
      end.to raise_error(%(schema failed for "/nullable" "get" "200" "application/json": The property '#/obj' of type object did not match any of the required schemas))
    end

    describe 'oneOf' do
      it 'passes #1' do
        expect do
          expect(validator.validate_response("GET", "/one_of", 200, "application/json", '{"image":{"image_type":"image","image_name":"redis"}}'))
        end.to_not raise_error
      end

      it 'passes #2' do
        expect do
          expect(validator.validate_response("GET", "/one_of", 200, "application/json", '{"image":{"image_type":"repo","image_repo":"uchiru/uchiru"}}'))
        end.to_not raise_error
      end

      it 'failed unknown type' do
        expect do
          expect(validator.validate_response("GET", "/one_of", 200, "application/json", '{"image":{"image_type":"some","image_repo":"uchiru/uchiru"}}'))
        end.to raise_error(%(schema failed for "/one_of" "get" "200" "application/json": The property '#/image' of type object did not match any of the required schemas))
      end

      it 'failed broken obj' do
        expect do
          expect(validator.validate_response("GET", "/one_of", 200, "application/json", '{"image":{"image_type":"image","image_repo":"uchiru/uchiru"}}'))
        end.to raise_error(%(schema failed for "/one_of" "get" "200" "application/json": The property '#/image' of type object did not match any of the required schemas))
      end
    end
  end
end


lib/strict_openapi3/parser.rb
module StrictOpenapi3
  class Parser
    PATH_NAME_RE = /^(\/(\{?[a-zA-Z0-9_]+\}?)?)+$/
    VERBS = ["get", "post", "patch", "put", "delete"]
    PATH_PARAM_SCHEMAS = [
      {"type" => "string"},
      {"type" => "integer"},
    ]
    QUERY_PARAM_SCHEMAS = [
      {"type" => "string"},
      {"type" => "integer"},
      {"type" => "boolean"},
    ]

    def initialize(json)
      @json = json
    end

    def to_schema!
      parse(@json)
    end

    private

    def parse(json)
      assert_keys("/", json, ["openapi", "info", "paths"], ["components", "tags"])
      out = {}
      parse_openapi(json["openapi"])
      parse_info(json["info"])
      parse_tags(json["tags"] || [])
      if json["components"]
        components = parse_components(json["components"])
      end
      out["paths"] = parse_paths(json["paths"], components, (json["tags"] || []).map { |t| t["name"] })
      out
    end

    def parse_openapi(value)
      assert_equal("/openapi", value, "3.0.0")
    end

    def parse_info(value)
      assert_keys("/info", value, ["version", "title"], ["description"])
    end

    def parse_tags(value)
      value.each_with_index do |tag, index|
        assert_keys("/tags[#{index}]", tag, ["name", "description"], [])
      end
    end

    def parse_paths(value, components, tags)
      paths = value.map do |name, path|
        assert_match("/paths:#{name}", name, PATH_NAME_RE)
        query = name[1..-1].split("/").map { |segment|
          if segment[0..0] == "{"
            {"kind" => "variable", "name" => segment[1..-2]}
          else
            {"kind" => "const", "value" => segment}
          end
        }
        {
          "orig" => name,
          "query" => query,
          "path" => parse_path("/paths:#{name}", path, components, query.select { |q| q["kind"] == "variable" }.map { |q| q["name"] }, tags)
        }
      end
      max_query_length = paths.map { |path| path["query"].length }.max.to_i
      paths.sort_by { |path|
        key = []
        max_query_length.times do |i|
          if path["query"][i].nil?
            key << 0
            key << "empty"
          elsif path["query"][i]["kind"] == "const"
            key << 1
            key << path["query"][i]["value"]
          elsif path["query"][i]["kind"] == "variable"
            key << 2
            key << path["query"][i]["name"]
          else
            raise "unreacheable point"
          end
        end
        key
      }
    end

    def parse_components(value)
      assert_keys("/components", value, ["schemas"], [])
      value["schemas"].map { |name, schema|
        [name, parse_schema("/components:#{name}", schema, {})]
      }.to_h
    end

    def parse_schema(prefix, value, components)
      unless value.is_a?(Hash)
        raise ParseError.new(%(Broken schema: "#{prefix}" should be hash)) 
      end
      assert_keys(prefix, value, [], ["type", "items", "required", "properties", "enum", "additionalProperties", "nullable", "pattern", "minLength", "$ref", "oneOf"])
      if !value.key?("type") && value.key?("$ref")
        assert_keys(prefix + "[no type]", value, [], ["$ref"])
        unless value["$ref"].index("#/components/schemas/") == 0
          raise ParseError.new(%(Broken schema: "#{prefix}:$ref" should have form "#/components/schemas/ComponentName"))
        end
        component_name = value["$ref"].sub("#/components/schemas/", "")
        if components.key?(component_name)
          JSON.load(JSON.dump(components[component_name]))
        else
          raise ParseError.new(%(Broken schema: "#{prefix}:$ref" unknown component "#{component_name}"))
        end
      elsif !value.key?("type") && value.key?("oneOf")
        assert_keys(prefix + "[no type]", value, [], ["oneOf"])
        {"oneOf" => value["oneOf"].each_with_index.map { |branch, i|
          parse_schema("#{prefix}:oneOf[#{i}]", branch, components)
        }}
      elsif value["type"] == "object"
        assert_keys(prefix + "[#{value["type"]}]", value, ["type", "properties"], ["required", "additionalProperties", "nullable"])
        out = {"type" => value["type"]}
        if value["required"]
          unless value["required"].is_a?(Array)
            raise ParseError.new(%(Broken schema: "#{prefix}:required" should be array)) 
          end
          if (value["required"] - value["properties"].keys).length > 0
            raise ParseError.new(%(Broken schema: "#{prefix}" "#{(value["required"] - value["properties"].keys).inspect}" are required but not descipted in properties section))
          end
          out["required"] = value["required"]
        end
        if value.key?("additionalProperties")
          if value["additionalProperties"] == true || value["additionalProperties"] == false
            out["additionalProperties"] = value["additionalProperties"]
          else
            out["additionalProperties"] = parse_schema("#{prefix}:additionalProperties", value["additionalProperties"], components)
          end
        else
          out["additionalProperties"] = false
        end
        out["properties"] = value["properties"].map do |prop, prop_spec|
          [prop, parse_schema("#{prefix}:#{prop}", prop_spec, components)]
        end.to_h
        if value["nullable"]
          {"oneOf" => [out, {"type" => "null"}]}
        else
          out
        end
      elsif value["type"] == "array"
        assert_keys(prefix + "[#{value["type"]}]", value, ["type", "items"], [])
        {
          "type" => "array",
          "items" => parse_schema("#{prefix}:items", value["items"], components)
        }
      elsif value["type"] == "string"
        assert_keys(prefix + "[#{value["type"]}]", value, ["type"], ["nullable", "pattern", "enum", "minLength"])
        if value.key?("enum")
          if value["enum"].is_a?(Array) && value["enum"].all? { |i| i.is_a?(String) }
            # its ok
          else
            raise ParseError.new(%(Broken schema: "#{prefix}:enum" should be array of strings")) 
          end
        end
        if value["nullable"]
          (value.keys - ["nullable"]).map { |k| 
            [k, value[k]]
          }.to_h.merge("type" => [value["type"], "null"])
        else
          value
        end
      elsif value["type"] == "integer" || value["type"] == "boolean" || value["type"] == "float"
        assert_keys(prefix + "[#{value["type"]}]", value, ["type"], ["nullable"])
        if value["nullable"]
          (value.keys - ["nullable"]).map { |k| 
            [k, value[k]]
          }.to_h.merge("type" => [value["type"], "null"])
        else
          value
        end
      else
        if value["type"]
          raise ParseError.new(%(Broken schema: "#{prefix}" unknown type "#{value["type"]}"))
        else
          raise ParseError.new(%(Broken schema: "#{prefix}" missed type))
        end
      end
    end

    def parse_path(prefix, value, components, path_params, tags)
      value.map do |name, method|
        assert_one_of(prefix, name, VERBS)
        assert_keys("#{prefix}:#{name}", method, ["responses", "summary", "tags"], ["description", "requestBody", "parameters"])
        assert_string_or_nil("#{prefix}:#{name}:description", method["description"])
        assert_string_or_nil("#{prefix}:#{name}:summary", method["summary"])
        unless method["tags"].is_a?(Array) && method["tags"].length == 1
          raise ParseError.new(%(Broken schema: "#{prefix}:#{name}:tags" should be array with one element))
        end
        unless tags.index(method["tags"][0])
          raise ParseError.new(%(Broken schema: "#{prefix}:#{name}:tags[#{method["tags"][0]}]" is not descripted in "tags" section of schema))
        end
        out = {}
        out["parameters"] = (method["parameters"] || []).each_with_index.map { |param, i|
          parse_param("#{prefix}:#{name}:parameters[#{i}]", param)
        }
        known_path_params = out["parameters"].select { |param| param["in"] == "path" }.map { |param| param["name"] }
        if (path_params - known_path_params).length > 0
          raise ParseError.new(%(Broken schema: "#{prefix}:#{name}:parameters" missed path params #{(path_params - known_path_params).inspect}))
        end
        if (known_path_params - path_params).length > 0
          raise ParseError.new(%(Broken schema: "#{prefix}:#{name}:parameters" unknown path params #{(known_path_params - path_params).inspect}))
        end
        if method.key?("requestBody")
          out["requestBody"] = parseRequestBody("#{prefix}:#{name}:requestBody", method["requestBody"], components)
        end
        out["responses"] = method["responses"].map { |code, response|
          assert_one_of("#{prefix}:#{name}:responses:#{code}", code.to_s, ["200", "201", "404", "409", "422", "default"])
          [code.to_s, parse_response("#{prefix}:#{name}:responses:#{code}", response, components)]
        }.to_h
        [name, out]
      end.to_h
    end

    def parse_response(prefix, response, components)
      assert_keys(prefix, response, ["description"], ["content"])
      out = {}
      if response.key?("content")
        out["content"] = response["content"].map { |content_type, content|
          assert_one_of("#{prefix}:#{content_type}", content_type, ["application/json", "text/plain"])
          if content_type == "application/json"
            assert_keys("#{prefix}:#{content_type}", content, ["schema"], [])
            [content_type, parse_schema("#{prefix}:#{content_type}:schema", content["schema"], components)]
          elsif content_type == "text/plain"
            assert_keys("#{prefix}:#{content_type}", content, ["example"], [])
            [content_type, {}]
          else
            raise "unreacheable point"
          end
        }.to_h
      end
      out
    end

    def parseRequestBody(prefix, value, components)
      assert_keys(prefix, value, ["content", "description", "required"], [])
      assert_one_of("#{prefix}:required", value["required"], [true, false])
      {
        "required" => value["required"],
        "content" => value["content"].map { |content_type, content|
          assert_one_of("#{prefix}:#{content_type}", content_type, ["application/json"])
          assert_keys("#{prefix}:#{content_type}", content, ["schema"], [])
          [content_type, parse_schema("#{prefix}:#{content_type}:schema", content["schema"], components)]
        }.to_h
      }
    end

    def parse_param(prefix, value)
      assert_keys(prefix, value, ["name", "in", "description", "required", "schema"], [])
      if value["in"] == "path"
        unless value["required"] == true
          raise ParseError.new(%(Broken schema: "#{prefix}:in[path]" required should be true))
        end
        assert_one_of("#{prefix}:in[schema]", value["schema"], PATH_PARAM_SCHEMAS)
      elsif value["in"] == "query"
        assert_one_of("#{prefix}:query[required]", value["required"], [true, false])
        assert_one_of("#{prefix}:query[schema]", value["schema"], QUERY_PARAM_SCHEMAS)
      else
        raise ParseError.new(%(Broken schema: "#{prefix}:in" unknown value "#{value["in"]}", allow to use ["path", "query"]))
      end
      value
    end

    def assert_match(prefix, value, reg)
      unless value =~ reg
        raise ParseError.new(%(Broken schema: "#{prefix}" expect match "#{reg}"))
      end
    end

    def assert_string_or_nil(prefix, value)
      if value.is_a?(String) || value.nil?
        # all ok
      else
        raise ParseError.new(%(Broken schema: "#{prefix}" expect nil or string, but "#{value}"))
      end
    end

    def assert_one_of(prefix, value, allowed = [])
      unless allowed.index(value)
        raise ParseError.new(%(Broken schema: "#{prefix}" unknown "#{value}", allow to use #{allowed.inspect})) 
      end
    end

    def assert_equal(prefix, value, expected)
      if value != expected
        raise ParseError.new(%(Broken schema: "#{prefix}" expect "#{expected}" but get "#{value}"))
      end
    end

    def assert_keys(prefix, hash, required = [], optional = [])
      if (required - hash.keys).length > 0
        raise ParseError.new(%(Broken schema: "#{prefix}" missed required keys #{(required - hash.keys).inspect}))
      end
      if (hash.keys - required - optional).length > 0
        raise ParseError.new(%(Broken schema: "#{prefix}" unknown keys #{(hash.keys - required - optional).inspect}))
      end
    end
  end
end
